pkg/template/processor_test.go

```diff
@@ -0,0 +1,604 @@
+		namespace, commit, user := templateVars()
+		values := paramsKeyValues(namespace, commit, user)
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+
+		// when
+		objs, err := p.Process(templateContent(namespaceObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		// then
+		require.NoError(t, err)
+		verifyNamespace(t, cl, namespace)
+	})
+
+	t.Run("should create role binding alone", func(t *testing.T) {
+		// given
+		s := addToScheme(t)
+		namespace, commit, user := templateVars()
+		values := paramsKeyValues(namespace, commit, user)
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+
+		// when
+		objs, err := p.Process(templateContent(roleBindingObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		// then
+		require.NoError(t, err)
+		verifyRoleBinding(t, cl, namespace)
+	})
+
+	t.Run("should create namespace and role binding", func(t *testing.T) {
+
+		t.Run("success", func(t *testing.T) {
+			// given
+			s := addToScheme(t)
+			namespace, commit, user := templateVars()
+			values := paramsKeyValues(namespace, commit, user)
+			cl := test.NewFakeClient(t)
+			p := template.NewProcessor(cl, s)
+
+			// when
+			objs, err := p.Process(templateContent(namespaceObj, roleBindingObj), values)
+			require.NoError(t, err)
+			err = p.Apply(objs, 2*time.Second)
+
+			// then
+			require.NoError(t, err)
+			verifyNamespace(t, cl, namespace)
+			verifyRoleBinding(t, cl, namespace)
+		})
+
+		t.Run("failure", func(t *testing.T) {
+
+			t.Run("client error", func(t *testing.T) {
+
+				t.Run("namespace does not exist", func(t *testing.T) {
+					// given
+					s := addToScheme(t)
+					namespace, commit, user := templateVars()
+					values := paramsKeyValues(namespace, commit, user)
+
+					cl := test.NewFakeClient(t)
+					cl.MockCreate = func(ctx context.Context, obj runtime.Object) error {
+						if obj.GetObjectKind().GroupVersionKind().Kind == "RoleBinding" {
+							return errors.New("mock error")
+						}
+						return nil
+					}
+					p := template.NewProcessor(cl, s)
+
+					// when
+					objs, err := p.Process(templateContent(namespaceObj, roleBindingObj), values)
+					require.NoError(t, err)
+					err = p.Apply(objs, 2*time.Second)
+
+					// then
+					require.Error(t, err)
+				})
+			})
+
+		})
+	})
+
+	t.Run("should update existing role binding", func(t *testing.T) {
+		// given
+		s := addToScheme(t)
+		namespace, commit, user := templateVars()
+		values := paramsKeyValues(namespace, commit, user)
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+
+		objs, err := p.Process(templateContent(roleBindingObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		require.NoError(t, err)
+		verifyRoleBinding(t, cl, namespace)
+
+		// when
+		objs, err = p.Process(templateContent(roleBindingObj, newUser), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
```

Comment:
correct me if I'm wrong - the RoleBinding that is processed/applied is the very same as the previous one, so there is no way you could verify that the object was really updated. You should change the object and then verify that the object was really changed.

Reply:
good point, let me do that

Reply:
actually, it was addressed in a previous commit: the updated rolebinding has an extra user:
```
objs, err = p.Process([]byte(namespaceAndRolebindingWithExtraUserTmpl), values)
require.NoError(t, err)
err = p.Apply(objs)
```